### PR TITLE
Fixed #1295 - 2.0: Setter Array, Dictionary, Parameter and  Document

### LINKED ISSUE #1295: 2.0: Setter Array, Dictionary, Parameter

### PR DESCRIPTION
**Pull request by hideki**
- Update setter methods of Array, Dictionary, Parameter and Document
- ReadOnlyArrayInterface and ReadOnlyDictionaryInterface no longer extends from Iterable. ReadOnlyDictionary and ReadOnlyArray implements Iterable
- Rename QueryRow to Result.
- Make ResultSet Iterable.
- Change Meta.documentID to Meta.id

### CODE REVIEW

File: shared/src/main/java/com/couchbase/lite/Array.java [hunk 1]
```diff
@@ -89,14 +89,69 @@ public Array set(int index, Object value) {
         return this;
     }
 
+    @Override
+    public ArrayInterface setString(int index, String value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Array.java]: I think it should return Array here.

File: shared/src/main/java/com/couchbase/lite/Array.java [hunk 2]
```diff
@@ -89,14 +89,69 @@ public Array set(int index, Object value) {
...
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setNumber(int index, Number value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Array.java]: I think it should return Array here.

File: shared/src/main/java/com/couchbase/lite/Array.java [hunk 3]
```diff
@@ -89,14 +89,69 @@ public Array set(int index, Object value) {
...
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setArray(int index, Array value) {
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setDictionary(int index, Dictionary value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Array.java]: I think it should return Array here.

File: shared/src/main/java/com/couchbase/lite/Array.java [hunk 4]
```diff
@@ -89,14 +89,69 @@ public Array set(int index, Object value) {
...
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setDate(int index, Date value) {
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setArray(int index, Array value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Array.java]: I think it should return Array here.

File: shared/src/main/java/com/couchbase/lite/Array.java [hunk 5]
```diff
@@ -89,14 +89,69 @@ public Array set(int index, Object value) {
...
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setBlob(int index, Blob value) {
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setDate(int index, Date value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Array.java]: I think it should return Array here.

File: shared/src/main/java/com/couchbase/lite/Array.java [hunk 6]
```diff
@@ -89,14 +89,69 @@ public Array set(int index, Object value) {
...
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setBoolean(int index, boolean value) {
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setBlob(int index, Blob value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Array.java]: I think it should return Array here.

File: shared/src/main/java/com/couchbase/lite/Array.java [hunk 7]
```diff
@@ -89,14 +89,69 @@ public Array set(int index, Object value) {
...
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setDouble(int index, double value) {
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setBoolean(int index, boolean value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Array.java]: I think it should return Array here.

File: shared/src/main/java/com/couchbase/lite/Array.java [hunk 8]
```diff
@@ -89,14 +89,69 @@ public Array set(int index, Object value) {
...
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setFloat(int index, float value) {
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setDouble(int index, double value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Array.java]: I think it should return Array here.

File: shared/src/main/java/com/couchbase/lite/Array.java [hunk 9]
```diff
@@ -89,14 +89,69 @@ public Array set(int index, Object value) {
...
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setLong(int index, long value) {
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setFloat(int index, float value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Array.java]: I think it should return Array here.

File: shared/src/main/java/com/couchbase/lite/Array.java [hunk 10]
```diff
@@ -89,14 +89,69 @@ public Array set(int index, Object value) {
...
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setInt(int index, int value) {
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setLong(int index, long value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Array.java]: I think it should return Array here.

File: shared/src/main/java/com/couchbase/lite/Array.java [hunk 11]
```diff
@@ -89,14 +89,69 @@ public Array set(int index, Object value) {
...
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setNumber(int index, Number value) {
+        return setObject(index, value);
+    }
+
+    @Override
+    public ArrayInterface setInt(int index, int value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Array.java]: I think it should return Array here.

File: shared/src/main/java/com/couchbase/lite/Array.java [hunk 12]
```diff
@@ -105,6 +160,61 @@ public Array add(Object value) {
         return this;
     }
 
+    @Override
+    public ArrayInterface addString(String value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Array.java]: I think it should return Array here.

File: shared/src/main/java/com/couchbase/lite/Array.java [hunk 13]
```diff
@@ -122,6 +232,61 @@ public Array insert(int index, Object value) {
         return this;
     }
 
+    @Override
+    public ArrayInterface insertString(int index, String value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Array.java]: I think it should return Array here.

File: shared/src/main/java/com/couchbase/lite/Dictionary.java [hunk 14]
```diff
@@ -125,6 +125,61 @@ public Dictionary set(String key, Object value) {
         return this;
     }
 
+    @Override
+    public DictionaryInterface setString(String key, String value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Dictionary.java]: I think it should return Dictionary here.

File: shared/src/main/java/com/couchbase/lite/Document.java [hunk 15]
```diff
@@ -137,11 +137,66 @@ public Document set(Map<String, Object> dictionary) {
...
     @Override
-    public Document set(String key, Object value) {
-        dictionary.set(key, value);
+    public Document setObject(String key, Object value) {
+        dictionary.setObject(key, value);
         return this;
     }
 
+    @Override
+    public DictionaryInterface setString(String key, String value) {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/Document.java]: I think it should return Document here. Same for others.

File: shared/src/main/java/com/couchbase/lite/FullTextResult.java [hunk 16]
```diff
@@ -22,7 +22,7 @@
 /**
  * A single result from a full-text Query.
  */
-public class FullTextQueryRow extends QueryRow {
+public class FullTextResult extends Result {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/FullTextResult.java]: Based on spec, we don't have a special result for FullTextResult.

Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/FullTextResult.java]: I will keep FullTextResult class till API is finalized.

Comment (hideki) [inline shared/src/main/java/com/couchbase/lite/FullTextResult.java]: https://github.com/couchbase/couchbase-lite-android/issues/1308

**Comment by hideki** [comment]
@pasin, your feedbacks applied to PR: https://github.com/couchbase/couchbase-lite-android/pull/1302

**Pull request merged**